README
NAME
    WWW::3172::Crawler::WebInterface - Provides a web frontend to
    WWW::3172::Crawler

VERSION
    version 0.001

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/WWW-3172-Crawler-WebInterface/>.

    The development version lives at
    <http://github.com/doherty/WWW-3172-Crawler-WebInterface> and may be
    cloned from
    <git://github.com/doherty/WWW-3172-Crawler-WebInterface.git>. Instead of
    sending patches, please fork this project using the standard git and
    github infrastructure.

SOURCE
    The development version is on github at
    <http://github.com/doherty/WWW-3172-Crawler-WebInterface> and may be
    cloned from <git://github.com/doherty/WWW-3172-Crawler-WebInterface.git>

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

0.001     2011-10-03
 - First stab at an interface to the crawler
    - Required stats: meta description and keywords
    - Extra stats: load time, size
    - Extra features: Obeys robots.txt, crawls from most linked-to to least
    - It's pretty!


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::Mock::LWP::Dispatch" => 0,
    "Test::More" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Provides a web frontend to WWW::3172::Crawler",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "WWW-3172-Crawler-WebInterface",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "WWW::3172::Crawler::WebInterface",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer" => 0,
    "Dancer::Plugin::DataFu" => 0,
    "Data::Table" => 0,
    "HTML::Entities" => 0,
    "List::UtilsBy" => 0,
    "WWW::3172::Crawler" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/app.pl"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


profiles/crawl.pl
use Data::Validate::URI qw(is_web_uri);

our $crawl = {
    url => {
        label       => 'URL to begin crawling:',
        error       => 'Invalid URL',
        required    => 1,
        value       => 'http://', # a hint for the user
        element     => { type => 'input_text' },
        validation  => sub {
            my ($form, $field, $params) = @_;
            $form->error($field, $field->{error}) unless is_web_uri($field->{value});
        },
        filters     => [qw(trim strip lowercase)],
    },
    max => {
        label       => 'Maximum number of pages to crawl:',
        required    => 0,
        value       => 5, # 50
        element     => { type => 'input_text' },
        validation  => sub {
            my ($form, $field, $params) = @_;
            $form->error($field, 'Cannot crawl more than 200 pages at once') unless $field->{value} < 200;
        },
        filters     => [qw(trim strip numeric)],
    },
};



lib/WWW/3172/Crawler/WebInterface.pm
package WWW::3172::Crawler::WebInterface;
use strict;
use warnings;

use Dancer ':syntax';
use Dancer::Plugin::DataFu;
use WWW::3172::Crawler;
use List::UtilsBy qw(nsort_by);
use HTML::Entities qw(encode_entities);
use Data::Table;
# ABSTRACT: Provides a web frontend to WWW::3172::Crawler
# VERSION

our $imported_crawler;
sub import {
    $imported_crawler = $_[1];
}

get '/' => sub {
    return template 'index.tt', { form => form->render('crawl', '/~doherty/crawler/submit_crawl', 'crawl.url', 'crawl.max') };
};

post '/submit_crawl' => sub {
    my $input = form;
    if ($input->validate('crawl.url', 'crawl.max')) {
        my $crawler = $imported_crawler || WWW::3172::Crawler->new(
            host    => $input->params->{'crawl.url'},
            max     => $input->params->{'crawl.max'},
        );
        my $crawl_data = $crawler->crawl;

        return template 'table.tt', {
            main_table  => _main_table($crawl_data),
            stats_table => _stats_table($crawl_data),
        };
    }
    redirect '/';
};

sub _main_table {
    my $crawl_data = shift;

    my $headers = ['URL', 'Keywords', 'Description', 'Size (bytes)', 'Speed (s)'];
    my @rows;
    while (my ($url, $data) = each %$crawl_data) {
        push @rows, [
            "<a href='$url'>" . encode_entities($url) . "</a>",
            $data->{keywords},
            $data->{description},
            $data->{size},
            sprintf("%.5f", $data->{speed}),
        ];
    }
    return Data::Table->new(\@rows, $headers, 0)->html;
}

sub _stats_table {
    my $crawl_data = shift;
    my %keywords;

    while (my ($url, $data) = each %$crawl_data) {
        my @these_keywords = split /,\s?|\s+/, ($data->{keywords} || '');
        $keywords{$_}++ for @these_keywords;
    }
    return '' if keys %keywords == 0;

    my $html = '<ul><li>';
    $html .= join '</li><li>', map { encode_entities($_) }
        nsort_by { $keywords{$_} } keys %keywords;
    $html .= '</li></ul>';

    return $html;
}

true;



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Provides a web frontend to WWW::3172::Crawler",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::Mock::LWP::Dispatch" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "WWW-3172-Crawler-WebInterface",
  "EXE_FILES" => [
    "bin/app.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "WWW::3172::Crawler::WebInterface",
  "PREREQ_PM" => {
    "Dancer" => 0,
    "Dancer::Plugin::DataFu" => 0,
    "Data::Table" => 0,
    "HTML::Entities" => 0,
    "List::UtilsBy" => 0,
    "WWW::3172::Crawler" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



